plutus-tx/src/Language/PlutusTx/Traversable.hs

```diff
@@ -0,0 +1,149 @@
+--    equivalent to traversal with a constant applicative functor
+--    ('foldMapDefault').
+--
+-- References:
+-- [1] The Essence of the Iterator Pattern, Jeremy Gibbons and Bruno C. d. S. Oliveira
+class (Functor t, Foldable t) => Traversable t where
+    -- | Map each element of a structure to an action, evaluate these actions
+    -- from left to right, and collect the results. For a version that ignores
+    -- the results see 'Language.PlutusTx.traverse_'.
+    traverse :: Applicative f => (a -> f b) -> t a -> f (t b)
+
+    -- All the other methods are deliberately omitted,
+    -- to make this a one-method class which has a simpler representation
+
+
+instance Traversable [] where
+    {-# INLINABLE traverse #-}
+    traverse f = go
+      where
+        go []     = pure []
+        go (x:xs) = liftA2 (:) (f x) (go xs)
+
+instance Traversable Maybe where
+    {-# INLINABLE traverse #-}
+    traverse _ Nothing  = pure Nothing
+    traverse f (Just a) = Just <$> f a
+
+instance Traversable (Either c) where
+    {-# INLINABLE traverse #-}
+    traverse _ (Left a)  = pure (Left a)
+    traverse f (Right a) = Right <$> f a
+
+instance Traversable ((,) c) where
+    {-# INLINABLE traverse #-}
+    traverse f (c, a) = (c,) <$> f a
+
+instance Traversable Identity where
+    {-# INLINABLE traverse #-}
+    traverse f (Identity a) = Identity <$> f a
+
+-- | Evaluate each action in the structure from left to right, and
+-- collect the results. For a version that ignores the results
+-- see 'Data.Foldable.sequenceA_'.
+sequenceA :: (Traversable t, Applicative f) => t (f a) -> f (t a)
+{-# INLINE sequenceA #-}  -- See Note [Inline default methods]
+sequenceA = traverse id
+
+-- | 'for' is 'traverse' with its arguments flipped. For a version
+-- that ignores the results see 'Data.Foldable.for_'.
+for :: (Traversable t, Applicative f) => t a -> (a -> f b) -> f (t b)
+{-# INLINE for #-}
+for = flip traverse
+
+-- | This function may be used as a value for `fmap` in a `Functor`
+--   instance, provided that 'traverse' is defined. (Using
+--   `fmapDefault` with a `Traversable` instance defined only by
+--   'sequenceA' will result in infinite recursion.)
+--
+-- @
+-- 'fmapDefault' f ≡ 'runIdentity' . 'traverse' ('Identity' . f)
+-- @
+fmapDefault :: forall t a b . Traversable t
+            => (a -> b) -> t a -> t b
+{-# INLINE fmapDefault #-}
+-- See Note [Function coercion] in Data.Functor.Utils.
+fmapDefault = coerce (traverse :: (a -> Identity b) -> t a -> Identity (t b))
+
+-- | This function may be used as a value for `Data.Foldable.foldMap`
+-- in a `Foldable` instance.
+--
+-- @
+-- 'foldMapDefault' f ≡ 'getConst' . 'traverse' ('Const' . f)
+-- @
+foldMapDefault :: forall t m a . (Traversable t, Monoid m)
+               => (a -> m) -> t a -> m
+{-# INLINE foldMapDefault #-}
+-- See Note [Function coercion] in Data.Functor.Utils.
+foldMapDefault = coerce (traverse :: (a -> Const m ()) -> t a -> Const m (t ()))
```

Comment:
I don't think `coerce` works with the plugin

Reply:
I think it may in fact work... I'm not sure if `Coercible` proper works. At least newtype coercions that you get from GND work.

Reply:
It does! I added a test case.